### ISSUE TITLE
When the checkpoint file reaches 100MB or more, it will be sharded into multiple files, but the reader class doesn't have the ability to parse it.

### ISSUE THREAD

**Issue by Caedin** [jeppe742/DeltaLakeReader#28]
I am using the reader for a project and it has worked for a few of my datasets, but I was getting back significantly less rows of data on my new dataset.

Upon investigation I found that when the checkpoint files reach 100MB they are sharded out into multiple files, but the reader has no logic for handling this.

This dataset works fine for us:
![image](https://user-images.githubusercontent.com/8877753/131736312-9370e6b4-2e52-402e-aa8f-476251cfd753.png)

But this one does not:
![image](https://user-images.githubusercontent.com/8877753/131736407-48f05609-ae41-4c2e-9824-81d770641537.png)

Upon investigating the code I believe this is where the bug is occuring:
![image](https://user-images.githubusercontent.com/8877753/131736595-4bd2b831-d3c9-49f4-940e-969c23aa15c8.png)

**Comment by jeppe742** [comment]
Hey @Caedin.

I wasn't aware of the checkpoint sharding, so thanks for bringing it to my knowledge.
Have actually never seen such large checkpoints. You must have a huge amount of files 😉 

Should be possible to read the checkpoints one at a time, so let me have a look at it

**Comment by jeppe742** [comment]
Hey @Caedin 

Do you have a rough idea of how many files you have in your table? (or just in one of the sharded checkpoints).
I'm trying to reproduce it, but having some small issues creating a table large enough to get this sharding. 
Even with 500.000 files the checkpoint is only 10MB :sweat_smile:

**Comment by jeppe742** [comment]
Hey @Caedin 

I have tried to create a fix in #29, but I might need your help to test it, since I haven't been able to reproduce such a large table ;)

**Comment by Caedin** [comment]
Thanks for taking a look at this Jeppe.

The dataset had 4.4 million files in the checkpoints. I am not sure what they are doing to create so many files, or why the json is so large. The 9 checkpoint files they had sum up to about 900MB of parquet for the 4.4 million files.

I took a look at the solution you proposed, and it should work. I cobbled something very similar together last Thursday and the code worked. 

I am still having an issue with the memory of the program due to such large checkpoints. I asked the team responsible for the dataset to optimize their table and they said they would. Hopefully that fixes the issue. Additionally, I am going to cache the DeltaTable object in my Azure Function, so that it only needs to read in the checkpoint list once every 15 minutes.
